app/lib/service/rate_limit/rate_limit.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:clock/clock.dart';
+import 'package:collection/collection.dart';
+import 'package:pub_dev/audit/backend.dart';
+
+import '../../account/agent.dart';
+import '../../audit/models.dart';
+import '../../shared/configuration.dart';
+import '../../shared/exceptions.dart';
+import '../../shared/redis_cache.dart';
+
+/// Verifies if the current package upload has a rate limit and throws
+/// if the limit has been exceeded.
+Future<void> verifyPackageUploadRateLimit({
+  required AuthenticatedAgent agent,
+  required String package,
+}) async {
+  final operation = AuditLogRecordKind.packagePublished;
+
+  if (agent is AuthenticatedUser) {
+    await _verifyRateLimit(
+      rateLimit: _getRateLimit(operation, RateLimitScope.user),
+      userId: agent.userId,
+    );
+  } else {
+    // apply per-user rate limit on non-user agents as they were package-specific limits
+    await _verifyRateLimit(
+      rateLimit: _getRateLimit(operation, RateLimitScope.user),
+      package: package,
+    );
+  }
+
+  // regular package-specific limits
+  await _verifyRateLimit(
+    rateLimit: _getRateLimit(operation, RateLimitScope.package),
+    package: package,
+  );
+}
+
+RateLimit? _getRateLimit(String operation, RateLimitScope scope) {
+  return activeConfiguration.rateLimits?.firstWhereOrNull(
+    (r) => r.operation == operation && r.scope == scope,
+  );
+}
+
+Future<void> _verifyRateLimit({
+  required RateLimit? rateLimit,
+  String? package,
+  String? userId,
+}) async {
+  if (rateLimit == null) {
+    return;
+  }
+  if (rateLimit.burst == null &&
+      rateLimit.hourly == null &&
+      rateLimit.daily == null) {
+    return;
+  }
+
+  final cacheKeyParts = [
+    rateLimit.operation,
+    rateLimit.scope.name,
+    if (package != null) 'package-$package',
+    if (userId != null) 'userId-$userId',
+  ];
+  final entryKey = Uri(pathSegments: cacheKeyParts).toString();
```

Comment:
```suggestion
  final cacheEntry = cache.rateLimitUntil([
    rateLimit.operation,
    rateLimit.scope.name,
    if (package != null) 'package-$package',
    if (userId != null) 'userId-$userId',
  ].join('/'));
  
  final current = await cacheEntry.get();
  if (current != null && current.isAfter(clock.now())) {
      throw RateLimitException(...
  }
```

Calling `auditBackend.getEntriesFromLastDay();` can incur some cost and pain, we should probably check the cache before we do an expensive operation.

I do see the problem that our cache doesn't contain the window or windowAsText for rate-limit that was applied.

**Proposal:** We make the `RateLimitException` serializable to json and put it in the cache :rofl:

Reply:
We can also leave this as TODO for later.